experimental/AlgebraicStatistics/src/DiscreteGraphicalModels.jl

```diff
@@ -0,0 +1,326 @@
+  DiscreteGraphicalModel(G, states, Dict{Symbol, VarName}(:t => t_var_name, :p => p_var_name))
+end
+
+varnames(M::DiscreteGraphicalModel) = M.varnames
+states(M::DiscreteGraphicalModel) = M.states
+n_states(M::DiscreteGraphicalModel) = length(states(M))
+maximal_cliques(M::DiscreteGraphicalModel) = maximal_cliques(graph(M))
+
+function Base.show(io::IO, M::DiscreteGraphicalModel{T, L}) where {T, L}
+  io = pretty(io)
+  if is_terse(io)
+    print(io, "Discrete Graphical Model on a")
+    !(L == Nothing) && print(io, " labelled")
+    print(io, " $T")
+  else
+    print(io, "Discrete Graphical Model on a $(graph(M))")
+  end
+  print(io, " with states ", string(M.states))
+end
+
+@doc raw"""
+    @attr Tuple{ModelRing, GenDict} model_ring(M::DiscreteGraphicalModel; cached=false)
+
+Return the ring in which the statistical model lives, together with a
+Dict for indexing its generators. This is the same for directed and for
+undirected graphs.
+
+``` jldoctest 
+julia> M = discrete_graphical_model(graph_from_edges([[1,2], [2,3]]), [2,2,2])
+Discrete Graphical Model on a Undirected graph with 3 nodes and 2 edges with states [2, 2, 2]
+
+julia> _, MR_gens = model_ring(M);
+
+julia> MR_gens[1, 2, 1]
+p[1,2,1]
+
+julia> MR_gens[(1, 2, 1)]
+p[1,2,1]
+```
+"""
+@attr Tuple{
+  ModelRing,
+  GenDict
+} function model_ring(M::DiscreteGraphicalModel; cached=false)
+  random_variables = 1:n_states(M)
+  state_spaces = [1:s for s in states(M)]
+  varindices = collect(Iterators.product(state_spaces...))
+
+  # TODO: the base ring should come from the model, leaving as QQ for now
+  return model_ring(QQ, varnames(M)[:p] => varindices)
+end
+
+function state_space(M::DiscreteGraphicalModel, C::Vector{Int})
+  return length(C) == 1 ? (1:states(M)[C[1]]) : Iterators.product([1:states(M)[i] for i in C]...)
+end
+
+@doc raw"""
+  @attr Tuple{MPolyRing, GenDict} parameter_ring(M::DiscreteGraphicalModel{Graph{Undirected}, T}; cached=false)
+
+Return the ring of parameters of the statistical model, together with a
+Dict for indexing its generators.
+
+``` jldoctest 
+julia> M = discrete_graphical_model(graph_from_edges([[1,2], [2,3]]), [2,2,2])
+Discrete Graphical Model on a Undirected graph with 3 nodes and 2 edges with states [2, 2, 2]
+
+julia> _, PR_gens = parameter_ring(M);
+
+julia> PR_gens[[1, 2], (2, 2)]
+t[1, 2](2, 2)
+```
+"""
+@attr Tuple{
+  MPolyRing,
+  GenDict
+} function parameter_ring(M::DiscreteGraphicalModel{Graph{Undirected}, T}; cached=false) where T
+  cliques = sort(sort.(collect.(maximal_cliques(graph(M)))))
+  Xs = [sort(vec(collect(state_space(M, C)))) for C in cliques]
+  params = [(C, x) for (C, X) in Iterators.zip(cliques, Xs) for x in X]
+  gen_names = sort([varnames(M)[:t] * string(C) * string(x) for (C, x) in params])
+  R, t = polynomial_ring(QQ, gen_names; cached=cached)
+  gens_dict = Dict((C, x) => t[i] for (i, (C, x)) in enumerate(params))
```

Comment:
This should do the same but shorter:
```suggestion
  gens_dict = Dict(zip(params, t))
```